### ISSUE TITLE
Stream race: default-stream consumer reads pre-copy state from async transfer-stream copy (gfx1101 / ROCm 7.2.2 / torch 2.11.0)

### ISSUE THREAD

**Issue by Dev-next-gen** [ROCmSoftwarePlatform/pytorch#3194]
## Summary

On a 5× AMD RX 7800 XT (gfx1101) system with ROCm 7.2.2 and PyTorch 2.11.0+rocm7.2.2 nightly, an async CPU→GPU `Tensor.copy_(..., non_blocking=True)` issued on a non-default `torch.cuda.Stream` is not implicitly synchronized with the default stream. A consumer op (`matmul`) that runs on the default stream immediately after the async copy reads pre-copy device memory in ~98% of iterations, producing numerically incorrect results.

Adding `default_stream.wait_stream(transfer_stream)` between the copy and the consumer op resolves the issue (0% mismatch over the same iteration counts).

The reproducer below uses **pure `torch` only** — no diffusers, no torchao, no `accelerate`, no high-level offload abstractions. Just two streams, a pinned-CPU tensor, an async copy, and a matmul.

## Environment

| | |
|---|---|
| OS | Linux 6.8.0-110-generic (Ubuntu) |
| Python | 3.12.3 |
| PyTorch | `2.11.0+rocm7.2.2.git4e323059` |
| `torch.version.hip` | `7.2.53211` |
| ROCm | 7.2.2 |
| GPUs | 5× AMD Radeon RX 7800 XT (gfx1101), 17.2 GB each |

## Reproducer

```python
"""
Minimal reproducer: stream race between async CPU->GPU copy and dependent matmul on AMD ROCm.

Compares three configurations:
  (1) Reference: blocking copy + matmul on default stream (correct).
  (2) Race:      async copy on transfer stream + immediate matmul on default stream
                 with NO explicit synchronization.
  (3) Fixed:     same as (2) but with `default_stream.wait_stream(transfer_stream)`
                 between the copy and the matmul.
"""

import argparse
import sys
import torch


def reference_result(A_cpu, B_gpu, device):
    A_gpu = A_cpu.to(device)  # blocking copy
    torch.cuda.synchronize(device)
    out = A_gpu @ B_gpu
    torch.cuda.synchronize(device)
    return out


def run_iter(A_cpu, B_gpu, device, transfer_stream, mode: str):
    size = A_cpu.shape[0]
    A_gpu = torch.empty(size, size, device=device, dtype=A_cpu.dtype)

    with torch.cuda.stream(transfer_stream):
        A_gpu.copy_(A_cpu, non_blocking=True)

    if mode == "fixed":
        torch.cuda.current_stream(device).wait_stream(transfer_stream)
    # mode == "race": no barrier on purpose

    out = A_gpu @ B_gpu
    torch.cuda.synchronize(device)
    return out


def count_mismatches(mode, A_cpu, B_gpu, device, transfer_stream, expected,
                     iters, rtol, atol):
    fails = 0
    for _ in range(iters):
        out = run_iter(A_cpu, B_gpu, device, transfer_stream, mode)
        if not torch.allclose(out, expected, rtol=rtol, atol=atol):
            fails += 1
    return fails


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--size", type=int, default=4096)
    p.add_argument("--iters", type=int, default=100)
    p.add_argument("--device-index", type=int, default=0)
    p.add_argument("--dtype", choices=["float32", "float16", "bfloat16"], default="float32")
    p.add_argument("--rtol", type=float, default=1e-4)
    p.add_argument("--atol", type=float, default=1e-4)
    p.add_argument("--seed", type=int, default=0)
    args = p.parse_args()

    torch.manual_seed(args.seed)
    device = torch.device("cuda", args.device_index)
    dtype = getattr(torch, args.dtype)

    print(f"torch={torch.__version__}  hip={getattr(torch.version, 'hip', None)}")
    props = torch.cuda.get_device_properties(args.device_index)
    print(f"device[{args.device_index}]={props.name}  arch={getattr(props, 'gcnArchName', '?')}")

    A_cpu = torch.randn(args.size, args.size, dtype=dtype, pin_memory=True)
    B_gpu = torch.randn(args.size, args.size, device=device, dtype=dtype)

    expected = reference_result(A_cpu, B_gpu, device)
    transfer_stream = torch.cuda.Stream(device=device)

    fails_race = count_mismatches("race", A_cpu, B_gpu, device, transfer_stream,
                                  expected, args.iters, args.rtol, args.atol)
    fails_fixed = count_mismatches("fixed", A_cpu, B_gpu, device, transfer_stream,
                                   expected, args.iters, args.rtol, args.atol)

    print(f"mode=race   mismatches: {fails_race} / {args.iters}")
    print(f"mode=fixed  mismatches: {fails_fixed} / {args.iters}")


if __name__ == "__main__":
    main()
```

Save as `repro_stream_race.py` and run:

```
python repro_stream_race.py --size 4096 --iters 100 --device-index 0
```

## Observed results

Across multiple configurations on this machine, mode=race produces incorrect results
in ~98% of iterations, while mode=fixed produces correct results in every iteration:

| Device | Size | Dtype | mode=race | mode=fixed |
|---|---|---|---|---|
| GPU 0 | 4096 | float32 | **99 / 100** | 0 / 100 |
| GPU 0 | 4096 | float16 | 49 / 50 | 0 / 50 |
| GPU 0 | 4096 | bfloat16 | 49 / 50 | 0 / 50 |
| GPU 1 | 4096 | float32 | 49 / 50 | 0 / 50 |
| GPU 4 | 4096 | float32 | 49 / 50 | 0 / 50 |
| GPU 0 | 1024 | float32 | 49 / 50 | 0 / 50 |

Across runs, exactly one iteration per run produces a correct result in mode=race
(49/50 or 99/100). I have not isolated which specific iteration this is — happy to
add per-iteration logging if useful. One hypothesis is that the first transfer
after stream creation benefits from a synchronization path that subsequent
transfers do not, but I have not verified this.

## Expected behavior

Either:
- (A) `Stream.copy_(..., non_blocking=True)` should produce a consumer-visible result on the default stream without requiring an explicit `wait_stream` (i.e., implicit ordering between the transfer stream and the default stream should hold), or
- (B) the API surface should make explicit that an explicit barrier is required between non-default-stream async transfers and default-stream consumers, regardless of backend.

On this configuration, neither holds: the race is silent and produces numerically incorrect results without any error, warning, or stream-validation diagnostic.

## Actual behavior

Default-stream consumer reads pre-copy device memory in ~98% of iterations across all
tested dtypes (fp32, fp16, bf16), all five `gfx1101` devices in the system, and
both small (1024) and medium (4096) tensor sizes.

## Workaround

`default_stream.wait_stream(transfer_stream)` between the async copy and the consumer
op resolves the race in every tested configuration. This is the workaround being
upstreamed at the application level in diffusers' `enable_group_offload(use_stream=True)`
code path: huggingface/diffusers#13502.

## What has NOT been tested

- CUDA hardware (this is an AMD-only system; cannot directly compare implicit-ordering behavior on NVIDIA).
- Other ROCm versions or torch builds beyond `2.11.0+rocm7.2.2.git4e323059`.
- Other AMD architectures (only `gfx1101` available on this system).
- Other transfer entry points than `Tensor.copy_(..., non_blocking=True)` — e.g., `Tensor.to(device, non_blocking=True)`, `aten._to_copy.default`. The diffusers PR linked above also exercises `tensor.to(device, non_blocking=True)` paths but I have not isolated whether the underlying race signature is identical.
- Other consumer ops than `matmul`. I have not checked whether ops with different scheduling characteristics (elementwise, conv, sdp_attention, etc.) reproduce the same race rate.

I can extend the reproducer along any of these axes if useful — please point me at what would be most valuable for the runtime team.

## Related

- huggingface/diffusers#13502 — application-level workaround in `ModuleGroup._onload_from_memory`. Approved by @jammm (AMD) on 2026-04-30; routed here for PyTorch/HIP-level investigation per @jammm's suggestion.
- https://github.com/Dev-next-gen/flux-amd-rocm — full-stack reproducer at the FLUX.1-dev / int8 / `enable_group_offload(use_stream=True)` level (original context where the bug surfaced).
- https://github.com/Dev-next-gen/flux-amd-rocm/blob/main/docs/bugs.md — Bug #3 in the project's bug log.

cc @jammm

**Comment by jeffdaily** [comment]
Thanks for the very clean reproducer — it's small, self-contained, and reproduces 1:1 on my end too (gfx90a / MI250X / ROCm 7.2.5 / torch nightly: 99/100 mismatches in `mode=race`). That made it quick to drill into.

The conclusion is that this is not a PyTorch or ROCm runtime bug — it's a cross-stream synchronization hazard that PyTorch's documented stream-semantics contract assigns to the user. The same code would race on NVIDIA CUDA; it's not architecture- or vendor-specific. Here's the evidence and how to verify it yourself.

## What the contract says

From [`docs/source/notes/cuda.rst`](https://github.com/pytorch/pytorch/blob/main/docs/source/notes/cuda.rst#cuda-streams) — the "CUDA streams" section:

> Operations inside each stream are serialized in the order they are created, but operations from different streams can execute concurrently in any relative order, unless explicit synchronization functions (such as `torch.cuda.synchronize` or `torch.cuda.Stream.wait_stream`) are used.

> When the "current stream" is the default stream, PyTorch automatically performs necessary synchronization when data is moved around. **However, when using non-default streams, it is the user's responsibility to ensure proper synchronization.**

The note even calls out an "incorrect" pattern that's structurally identical to the reproducer (`A = torch.empty(...).normal_()` on default, then `with torch.cuda.stream(s): B = torch.sum(A)`) and prescribes the fix: `s.wait_stream(default_stream)` *before* the side-stream op, plus `A.record_stream(s)` to keep the storage alive.

On the implementation side, `aten/src/ATen/native/cuda/Copy.cu` for CPU→CUDA `non_blocking=True` issues `cudaMemcpyAsync` on the *current* stream and only records an event against the pinned-host allocator (so the host staging buffer isn't reused mid-DMA). It deliberately does not insert any `record_stream`/event on the destination GPU tensor — there's no implicit cross-stream ordering, by design.

## How to verify with CSAN

The strongest evidence is from PyTorch's [CUDA Sanitizer (CSAN)](https://docs.pytorch.org/docs/stable/cuda._sanitizer.html), a `TorchDispatchMode` that tracks per-tensor read/write streams and raises `CUDASanitizerErrors` whenever a read on stream B is not ordered after a write on stream A. It works on ROCm — no extra setup, just an env var:

```
TORCH_CUDA_SANITIZER=1 python repro_stream_race.py --size 1024 --iters 3 --device-index 0
```

Running your exact reproducer this way produces (excerpted):

```
CSAN detected a possible data race on tensor with data pointer ...
Access by stream <transfer_stream> during kernel:
aten::copy_(Tensor(a!) self, Tensor src, bool non_blocking=False) -> Tensor(a!)
  ... A_gpu.copy_(A_cpu, non_blocking=True)
Previous access by stream 0 during kernel:
aten::empty.memory_format(...) -> Tensor
  ... A_gpu = torch.empty(size, size, device=device, dtype=A_cpu.dtype)
```

That's actually the *first* hazard CSAN sees — between the default-stream `aten::empty` and the side-stream `aten::copy_`. To also surface the matmul-vs-copy hazard (the one that produces the visible numerical mismatch you're seeing), you can fix the empty-vs-copy edge first by adding `transfer_stream.wait_stream(default_stream)` *before* the copy. Re-running CSAN then yields:

```
CSAN detected a possible data race on tensor with data pointer ...
Access by stream 0 during kernel:
aten::mm(Tensor self, Tensor mat2) -> Tensor
  ... out = A_gpu @ B_gpu
Previous access by stream <transfer_stream> during kernel:
aten::copy_(Tensor(a!) self, Tensor src, bool non_blocking=False) -> Tensor(a!)
  ... A_gpu.copy_(A_cpu, non_blocking=True)
```

So there are *two* distinct cross-stream hazards in the reproducer — both of which the docs' synchronization pattern fixes.

## A subtle point about your "fixed" form

The workaround in the issue (`default_stream.wait_stream(transfer_stream)` between the copy and the consumer) does eliminate the visible numerical mismatch — the matmul-vs-copy edge is now safe. But CSAN still flags the empty-vs-copy edge in that version, because the pre-copy `transfer_stream.wait_stream(default_stream)` is missing. For *this* particular loop the missing pre-edge happens not to bite (each iteration's `torch.empty` lands on a fresh block with no in-flight kernels), but it absolutely will if you reuse an `A_gpu` storage across iterations or if other default-stream work is in flight when the side-stream copy is issued — the CUDA caching allocator can hand the same block back with pending side-stream work, which is exactly the case the docs' "STILL REQUIRED!" note (see the second example in the streams section) is warning about.

The CSAN-clean pattern, matching the docs:

```python
A_gpu = torch.empty(...)
transfer_stream.wait_stream(default_stream)            # pre-copy
with torch.cuda.stream(transfer_stream):
    A_gpu.copy_(A_cpu, non_blocking=True)
A_gpu.record_stream(transfer_stream)                   # keep alive while DMA in flight
default_stream.wait_stream(transfer_stream)            # before consumer
out = A_gpu @ B_gpu
```

This passes CSAN cleanly and produces 0 mismatches.

## Why the visible mismatch rate is ~98%, not 100%

In `mode=race`, the matmul launch is queued onto the default stream while the DMA is still in flight on the transfer stream. The matmul issue path is short and the DMA is large — the matmul almost always wins the race. The handful of "correct" iterations are runs where, by scheduling chance, the matmul kernel doesn't start executing until after the DMA has completed. That's consistent with your observation that exactly one iteration per run is correct (often the first, when the streams are empty and the runtime ends up serializing by accident).

## Recommendation

I'd suggest closing this as expected stream semantics rather than a runtime fix. Two follow-ups that might still be valuable:

1. A docs-visibility tweak — referencing the streams note from the `Tensor.to(..., non_blocking=True)` and `Tensor.copy_(..., non_blocking=True)` docstrings, since users hitting this rarely find the streams note on their own. Happy to file that as a separate PR if there's interest.
2. **Use CSAN as your default tool any time you suspect a cross-stream race** — in this repo, in `diffusers`'s offload paths, in any other downstream stack you're poking. It's the fastest way to turn "98% wrong matmul output" into a precise stack trace that names the two unsynchronized streams and the tensors involved. It works on both CUDA and ROCm and is essentially zero-config (`TORCH_CUDA_SANITIZER=1`). The same investigation here would have been a one-shot CSAN run if applied to the diffusers reproducer at https://github.com/huggingface/diffusers/pull/13502 — both the missing pre-copy and post-copy edges would have been pointed out automatically.

Thanks again for the reproducer — even though it turned out to be a contract issue, it's the kind of report that's quick to triage exactly because it's minimal and self-contained.

**Comment by Dev-next-gen** [comment]
@jeffdaily thanks for the detailed walkthrough — and especially for re-running CSAN to surface both edges. Good catch on the missing pre-copy `transfer_stream.wait_stream(default_stream)` and `record_stream` in the workaround; agreed it would bite as soon as the allocator hands back a block with pending side-stream work, even if this particular loop happens to dodge it.

Happy to take the docs-visibility PR — I'll reference the streams note from the `Tensor.to(..., non_blocking=True)` and `Tensor.copy_(..., non_blocking=True)` docstrings against pytorch/pytorch. I'll also re-run CSAN against the diffusers offload path in huggingface/diffusers#13502 to confirm the same pre/post-copy edges and tighten the workaround there.

Feel free to close this whenever — I'll link the docs PR back here when it's up.

**Comment by Dev-next-gen** [comment]
Docs-visibility PR is up: pytorch/pytorch#182051 — adds a `.. note::` to `Tensor.to(non_blocking=True)` and `Tensor.copy_(non_blocking=True)` cross-referencing the streams note via a new `:ref:` label, no content duplication. cc @jeffdaily.
